toolkit/tools/grapher/grapher.go

```diff
@@ -22,12 +23,27 @@ var (
 	input  = exe.InputFlag(app, "Input json listing all local SRPMs")
 	output = exe.OutputFlag(app, "Output file to export the graph to")
 
-	logFile          = exe.LogFileFlag(app)
-	logLevel         = exe.LogLevelFlag(app)
-	profFlags        = exe.SetupProfileFlags(app)
-	strictGoals      = app.Flag("strict-goals", "Don't allow missing goal packages").Bool()
-	strictUnresolved = app.Flag("strict-unresolved", "Don't allow missing unresolved packages").Bool()
-	timestampFile    = app.Flag("timestamp-file", "File that stores timestamps for this program.").String()
+	logFile               = exe.LogFileFlag(app)
+	logLevel              = exe.LogLevelFlag(app)
+	profFlags             = exe.SetupProfileFlags(app)
+	strictGoals           = app.Flag("strict-goals", "Don't allow missing goal packages").Bool()
+	strictUnresolved      = app.Flag("strict-unresolved", "Don't allow missing unresolved packages").Bool()
+	timestampFile         = app.Flag("timestamp-file", "File that stores timestamps for this program.").String()
+	usePMCtoResolveCycles = app.Flag("usePMCtoresolvecycles", "Cycles will be resolved by downloading rpm packages from PMC if locally unavailable").Bool()
+	tlsClientCert         = app.Flag("tls-cert", "TLS client certificate to use when downloading files.").String()
+	tlsClientKey          = app.Flag("tls-key", "TLS client key to use when downloading files.").String()
+	packageURLlist        = app.Flag("packageURLlist", "PACKAGE_URL_LIST").Strings()
+
+	resolveCyclesFromUpstream     = app.Flag("resolve-cycles-from-upstream", "Let grapher resolve cycles by marking rpms available in repo as remote").Bool()
+	outDir                        = exe.OutputDirFlag(app, "Directory to download packages into.")
+	existingRpmsDir               = app.Flag("rpm-dir", "Directory that contains already built RPMs. Should contain top level directories for architecture.").Required().ExistingDir()
+	existingToolchainRpmDir       = app.Flag("toolchain-rpms-dir", "Directory that contains already built toolchain RPMs. Should contain top level directories for architecture.").Required().ExistingDir()
+	tmpDir                        = app.Flag("tmp-dir", "Directory to store temporary files while downloading.").String()
+	workerTar                     = app.Flag("tdnf-worker", "Full path to worker_chroot.tar.gz").Required().ExistingFile()
+	repoFiles                     = app.Flag("repo-file", "Full path to a repo file").Required().ExistingFiles()
+	usePreviewRepo                = app.Flag("use-preview-repo", "Pull packages from the upstream preview repo").Bool()
+	toolchainManifest             = app.Flag("toolchain-manifest", "Path to a list of RPMs which are created by the toolchain. Will mark RPMs from this list as prebuilt.").ExistingFile()
+	ignoreVersionToResolveSelfDep = app.Flag("ignore-version-to-resolve-selfdep", "Ignore package version while downloading package from upstream when resolving cycle").Bool()
```

Comment:
Please correct me if I'm wrong. My understanding is we only need to use `ignoreVersionToResolveSelfDep` for cases like `BuildRequires: ghc = %{version}-%{release}` inside `ghc.spec` itself. If there is no other scenario when we'd need this flag, then I think we should remove it entirely, since a build dependency on the current version of yourself is bug in the spec, which should be fixed by switching to `BuildRequires: ghc < %{version}-%{release}`. Otherwise, we're just masking an issue in the spec at an extra cost of the tooling's complexity.

Reply:
This is not a bug in the SPEC file. Mariner build-toolkit is written in such a way.

If there is a BR in `ghc.spec` as `BuildRequires = ghc`, `grapher` creates an edge from `BUILD-ghc` to `RUN-ghc` which can never be resolved. We're fixing it in following way,
* Clone the `RUN` node into a `REMOTE-UNRESOLVED` node and replace the edge with `BUILD-ghc` --> `REMOTE-UNRESOLVED-ghc` to edge.
* This `REMOTE-UNRESOLVED` node should be fetched by `graphPkgFetcher`
* But while cloning the version of the `RUN` node, its version is copied into the `REMOTE-UNRESOLVED` node as well. And `graphPkgFetcher` fires a `tdnf whatprovides` call with the version string attached. So, the relevant package will never be downloaded unless `graphPkgFetecher` ignores the version string.

Ideally, we should fix the `graphPkgFetcher` to use version string only if there is an explicit version dependency mentioned in the SPEC file. But that requires changes in the `PkgNode` itself which is used in multiple other places too.

Fixing every self-dependency with `<` is not a clean approach. Debian and Fedora are able to handle it naturally with explicit `<` versioning.

Reply:
Ah, you're right, @0xba1a!

I didn't think about the versionless `BuildRequires: ghc`. Failure to handle this correctly is a bug in our tooling, like you said, and I see why we need `ignoreVersionToResolveSelfDep` as a workaround. Only `BuildRequires: %{name} = %{version}-%{release}` is an invalid dependency in a spec.

I've created [a bug](https://microsoft.visualstudio.com/OS/_workitems/edit/45846176) to fix the issue with handling versionless dependencies. Once it's fixed, we should be able to remove `ignoreVersionToResolveSelfDep` but of that is out of scope for this PR.